File: Chohyeon Kim Elly kim/2023-05-16-Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        def subhelper(rt, sub):
            
            if not rt and not sub:
                return True

            if rt and sub and rt.val == sub.val:
                 return subhelper(rt.left, sub.left) and subhelper(rt.right,sub.right)
            else:
                return False
        
        if not root:
            return False

        res = False

        if root and subRoot and root.val == subRoot.val:
            res = subhelper(root,subRoot)

        
        return res or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: kiwoon-jung/2023-05-16-Subtree-of-Another-Tree.js
var isSubtree = function (root, subRoot) {
  if (subRoot && !root) return false;

  if (isMatch(root, subRoot)) return true;

  return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
};

function isMatch(r, s) {
  if (r && s) {
    return (
      r.val === s.val && isMatch(r.left, s.left) && isMatch(r.right, s.right)
    );
  }

  if (!r && s) return false;
  if (r && !s) return false;
  if (!r && !s) return true;
}
